### ISSUE TITLE
Stateful LSTM does not work with mixed precision

### ISSUE THREAD

**Issue by xxviktor** [keras-team/keras#14047]
**System information**  
- Have I written custom code (as opposed to using example directory):  yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04):  tested on Linux Ubuntu 20.04 and Windows 10
- TensorFlow backend (yes / no):  yes
- TensorFlow version:  '2.1.0'
- Keras version:  '2.2.4-tf'
- Python version:  3.7.7
- CUDA/cuDNN version:  11.0
- GPU model and memory:  RTX 2070, 8GB

**Describe the current behavior** 
Stateful LSTM does not work in mixed precision. When trying to run the code without casting:
```
LSTM(units=16, stateful=True)
```
I'm getting error:
```
Input 'b' of 'MatMul' Op has type float16 that does not match type float32 of argument 'a'.
```
Code could be fixed by manual cast to float32:
```
LSTM(units=16, stateful=True, dtype="float32")
```
But than it looks like Tensor cores are not used.

**Describe the expected behavior**  
Code should not throw errors.

**Code to reproduce the issue**  
```
import numpy as np
import tensorflow as tf
from tensorflow.keras import Model
from tensorflow.keras.layers import Input, Dense, LSTM
from tensorflow.keras.mixed_precision import experimental as mixed_precision
from tensorflow.keras.optimizers import Adam


gpus = tf.config.experimental.list_physical_devices('GPU')
if gpus:
    try:
        for gpu in gpus:
            tf.config.experimental.set_memory_growth(gpu, True)
            logical_gpus = tf.config.experimental.list_logical_devices('GPU')
    except RuntimeError as e:
        print(e)

policy = mixed_precision.Policy('mixed_float16')
mixed_precision.set_policy(policy)

X = np.random.normal(size=64).reshape(8,8,1)
Y = np.random.normal(size=8)

xinput = Input(batch_shape=(1, 8, 1))
x = Dense(16)(xinput)
x = LSTM(units=16, stateful=True)(x)
xoutput = Dense(1)(x)

train_model = Model(inputs=xinput, outputs=xoutput)
train_model.compile(optimizer=Adam(), loss='mean_squared_error')
train_model.summary()

train_model.fit(x=X, y=Y, epochs=1, shuffle=False)
```

**Comment by unsatcore** [comment]
Same thing with tensorflow

-     Have I written custom code (as opposed to using example directory): yes (e.g. the above example)
-     OS Platform and Distribution (e.g., Linux Ubuntu 16.04): NixOS 20.03
-     TensorFlow version: '2.3.0'
-     Python version: 3.7.7
-     CUDA/cuDNN version: 10.1
-     GPU model and memory: GTX 1070, 8GB

Perhaps, should be filed into tensorflow issues too?

**Comment by asaf92** [comment]
I experience the same thing when using GRU with `stateful=True`

Minimal example:
```
import tensorflow as tf
from tensorflow.keras.layers import Embedding, GRU, Dense
from tensorflow.keras.models import Sequential
from tensorflow.keras.mixed_precision import experimental as mixed_precision
policy = mixed_precision.Policy('mixed_float16')
mixed_precision.set_policy(policy)

model = Sequential([
    Embedding(128, 1024, batch_input_shape=[64, None]),
    GRU(1024, return_sequences=True, stateful=True),
    Dense(1024)
])
```

The error isn't raised when `stateful=True` is missing

**Comment by davzaman** [comment]
Same situation as @unsatcore 

- Wrote custom code, and get the matmul exception when using mixed precision. This happens on multiple different machines (different GPUs but same versioning).
- OS Ubuntu 18.04.5
- TF 2.4.1
- Python 3.8.3
- CUDA 11.0
- GPUs tried:
  - Tesla P40 24GB
  - V100 NVIDIA 32GB
  - A100 NVIDIA 40GB

**Comment by relativeflux** [comment]
I raised this in a [new issue](https://github.com/keras-team/keras/issues/15140), and it has been fixed. The cause, as I suspected, was incorrect dtype for the initial state tensor.

But the fix seems not to be present in TF 2.6.0... I'm still seeing the same error. I don't know how the TF/Keras teams integrate bug fixes, will this only be available in the next (presumably minor) release, or is it present in tf-nightly right now?
